mypy/checkexpr.py

```diff
@@ -176,6 +181,86 @@ def visit_call_expr(self, e: CallExpr) -> Type:
+                return self.check_typeddict_call_with_dict(callee, unique_arg, context)
+            if isinstance(unique_arg, CallExpr) and isinstance(unique_arg.analyzed, DictExpr):
+                # ex: Point(dict(x=42, y=1337))
+                return self.check_typeddict_call_with_dict(callee, unique_arg.analyzed, context)
+
+        if len(args) == 0:
+            # ex: EmptyDict()
+            return self.check_typeddict_call_with_kwargs(
+                callee, OrderedDict(), context)
+
+        self.chk.fail(messages.INVALID_TYPEDDICT_ARGS, context)
+        return AnyType()
+
+    def check_typeddict_call_with_dict(self, callee: TypedDictType,
+                                       kwargs: DictExpr,
+                                       context: Context) -> Type:
+        item_name_exprs = [item[0] for item in kwargs.items]
+        item_args = [item[1] for item in kwargs.items]
+
+        item_names = []  # List[str]
+        for item_name_expr in item_name_exprs:
+            if not isinstance(item_name_expr, StrExpr):
+                self.chk.fail(messages.TYPEDDICT_ITEM_NAME_MUST_BE_STRING_LITERAL, item_name_expr)
+                return AnyType()
+            item_names.append(item_name_expr.value)
+
+        return self.check_typeddict_call_with_kwargs(
+            callee, OrderedDict(zip(item_names, item_args)), context)
+
+    def check_typeddict_call_with_kwargs(self, callee: TypedDictType,
+                                         kwargs: 'OrderedDict[str, Expression]',
+                                         context: Context) -> Type:
+        if callee.items.keys() != kwargs.keys():
+            callee_item_names = callee.items.keys()
+            kwargs_item_names = kwargs.keys()
+            missing_item_names = [k for k in callee_item_names if k not in kwargs_item_names]
+            extra_item_names = [k for k in kwargs_item_names if k not in callee_item_names]
+
+            self.chk.fail(
+                'Expected items {} but found {}. Missing {}. Extra {}.'.format(
+                    list(callee_item_names),
+                    list(kwargs_item_names),
+                    missing_item_names,
+                    extra_item_names
+                ),
+                context)
+            return AnyType()
+
+        items = OrderedDict()  # type: OrderedDict[str, Type]
+        for (item_name, item_expected_type) in callee.items.items():
+            item_value = kwargs[item_name]
+
+            item_actual_type = self.chk.check_simple_assignment(
+                lvalue_type=item_expected_type, rvalue=item_value, context=item_value,
+                msg=messages.INCOMPATIBLE_TYPES,
+                lvalue_name='TypedDict item',
+                rvalue_name='expression')
+            items[item_name] = item_actual_type
+
+        mapping_value_type = join.join_type_list(list(items.values()))
+        fallback = self.chk.named_generic_type('typing.Mapping',
```

Comment:
The question about the fallback type is actually kind of tricky. It's not clear to me if `Mapping` is the most practical fallback definition, though I think that it's a type-safe one. 

In existing code, people likely use `Dict[str, Any]` for things that would be suitable for `TypedDict`. Introducing `TypedDict` gradually to such a codebase would likely imply getting rid of many of the `Dict[str, Any]` types or replacing them with `Mapping[str, Any]` (or adding casts), so the introduction wouldn't be perfectly gradual.

Here are alternatives that may or may not be reasonable:
* If a typed dict has uniform value types, make the fallback `Dict[str, value_type]`. This wouldn't be quite safe since `dict` values support `__del__` and adding new keys.
* Have two fallbacks (the second one could be special cased somehow), `Mapping[str, ...]` (as currently) and `Dict[str, Any]`. Code dealing with `Any` types won't be safe anyway, so this would still be safe for fully typed code but would also perhaps provide a smoother gradual typing story.

We don't need to decide this now -- it may well take some practical experimentation with real code to determine the best approach.